inference-engine/tests/ie_test_utils/common_test_utils/ngraph_test_utils.cpp

```diff
@@ -83,6 +85,367 @@ std::string name(const Node& n) {
     return n->get_friendly_name();
 }
 
+using AttrName = std::string;
+
+class Result {
+public:
+    explicit Result(std::string m = {}) : m_message(std::move(m)) {}
+
+    const std::string& message() const {
+        return m_message;
+    }
+
+    explicit operator bool() const {
+        return m_message.empty();
+    }
+
+    Result& operator+=(const std::string& msg) {
+        m_message.append(1, '\n').append(msg);
+        return *this;
+    }
+
+private:
+    std::string m_message;
+};
+
+class MemoryChunk {
+public:
+    using Data = std::vector<unsigned char>;
+    MemoryChunk(Data data) : m_data{std::move(data)} {}
+
+    Data::const_pointer data() const {
+        return m_data.data();
+    }
+
+    size_t size() const {
+        return m_data.size();
+    }
+
+private:
+    Data m_data;
+};
+
+template <typename AttrValue>
+class AttributeStorage {
+public:
+    bool insert_value(AttrName name, AttrValue value) {
+        return m_attributes.insert({std::move(name), std::move(value)}).second;
+    }
+
+    const AttrValue* get_value(const AttrName& name) const {
+        const auto found = m_attributes.find(name);
+        if (found != end(m_attributes)) {
+            return std::addressof(found->second);
+        }
+        return {};
+    }
+
+    std::size_t get_attributes_number() const {
+        return m_attributes.size();
+    }
+
+private:
+    std::map<AttrName, AttrValue> m_attributes;
+};
+
+class Storage : public AttributeStorage<MemoryChunk>,
+                public AttributeStorage<bool>,
```

Comment:
I think will be better to use virtual inheritance.

Reply:
@ilyachur , please convince me.
My arguments to stay as is:
- `Storage` is not espouse in header
- `Storage` is local helper class
- virtual interface bring indirect calls
- I'm not sure what should be virtual here
- whole `std` library works similar it has nonvirtual helper class

If you afraid of memory leaks (don't see any other risk here which might force to move to virtual interface) I can switch to private visibility of inherited classes:
```
class Storage : private AttributeStorage<MemoryChunk>,
                private AttributeStorage<bool>,
                private AttributeStorage<std::string>,
                private AttributeStorage<int8_t>,
                private AttributeStorage<int16_t>,
                private AttributeStorage<int32_t>,
                private AttributeStorage<int64_t>,
                private AttributeStorage<uint8_t>,
                private AttributeStorage<uint16_t>,
                private AttributeStorage<uint32_t>,
                private AttributeStorage<uint64_t>,
                private AttributeStorage<float>,
                private AttributeStorage<double>,
                private AttributeStorage<std::vector<int8_t>>,
                private AttributeStorage<std::vector<int16_t>>,
                private AttributeStorage<std::vector<int32_t>>,
                private AttributeStorage<std::vector<int64_t>>,
                private AttributeStorage<std::vector<uint8_t>>,
                private AttributeStorage<std::vector<uint16_t>>,
                private AttributeStorage<std::vector<uint32_t>>,
                private AttributeStorage<std::vector<uint64_t>>,
                private AttributeStorage<std::vector<float>>,
                private AttributeStorage<std::vector<double>>,
                private AttributeStorage<std::vector<std::string>> {
public:
    template <typename AttrValue>
    const AttributeStorage<AttrValue>* storage() const {
        return static_cast<const AttributeStorage<AttrValue>*>(this);
    }
    template <typename AttrValue>
    AttributeStorage<AttrValue>* storage() {
        return static_cast<AttributeStorage<AttrValue>*>(this);
    }
};
```

Reply:
Yeah, I see that `Storage` class is located in the cpp, and I asked about virtual inheritance only in order to avoid any problems with memory leaks when we destroy the object (because we already faced with such issues).

Reply:
done